src/community/features-templating/features-templating-core/src/test/java/org/geoserver/featurestemplating/builders/visitors/SelectionTemplateVisitorTest.java

```diff
@@ -0,0 +1,126 @@
+import org.geoserver.featurestemplating.builders.selectionwrappers.PropertySelectionWrapper;
+import org.geoserver.featurestemplating.readers.JSONTemplateReader;
+import org.geoserver.featurestemplating.readers.RecursiveJSONParser;
+import org.geoserver.featurestemplating.readers.TemplateReaderConfiguration;
+import org.geoserver.platform.resource.FileSystemResourceStore;
+import org.geoserver.platform.resource.Resource;
+import org.geotools.feature.simple.SimpleFeatureTypeBuilder;
+import org.geotools.jdbc.JDBCDataStore;
+import org.junit.Before;
+import org.junit.Test;
+import org.locationtech.jts.geom.Geometry;
+import org.opengis.feature.simple.SimpleFeatureType;
+import org.opengis.filter.expression.Expression;
+import org.opengis.filter.expression.Literal;
+import org.xml.sax.helpers.NamespaceSupport;
+
+public class SelectionTemplateVisitorTest {
+    private SimpleFeatureType simpleFeatureType;
+    private FileSystemResourceStore store;
+
+    @Before
+    public void setup() {
+        store =
+                new FileSystemResourceStore(
+                        new File(
+                                "src/test/resources/org/geoserver/featurestemplating/builders/visitors"));
+        SimpleFeatureTypeBuilder tb = new SimpleFeatureTypeBuilder();
+        tb.add("geometry", Geometry.class);
+        tb.add("one", String.class);
+        tb.add("b", String.class);
+        tb.add("c", String.class);
+        tb.setName("testFeatureType");
+        SimpleFeatureType schema = tb.buildFeatureType();
+        schema.getDescriptor("b").getUserData().put(JDBCDataStore.JDBC_NATIVE_TYPENAME, "json");
+        schema.getDescriptor("c").getUserData().put(JDBCDataStore.JDBC_NATIVE_TYPENAME, "json");
+        this.simpleFeatureType = schema;
+    }
+
+    @Test
+    public void testPropertySelectionVisitor() throws IOException {
+        PropertySelectionHandler propertySelectionHandler = handler();
+        RootBuilder root = getBuilderTree("testTemplate.json");
+        PropertySelectionVisitor visitor =
+                new PropertySelectionVisitor(propertySelectionHandler, simpleFeatureType);
+        root = (RootBuilder) root.accept(visitor, null);
+        List<TemplateBuilder> children = root.getChildren();
+        TemplateBuilder builder = children.get(0);
+        builder = builder.getChildren().get(0);
+        for (TemplateBuilder child : builder.getChildren()) {
+            AbstractTemplateBuilder abstractBuilder = (AbstractTemplateBuilder) child;
+            String key = abstractBuilder.getKey(null);
+            if ("b".equals(key)) {
+                assertTrue(child instanceof DynamicPropertySelection);
+            } else if ("three".equals(key)) {
+                TemplateBuilder dynamicInclude = abstractBuilder.getChildren().get(1);
+                assertTrue(dynamicInclude instanceof DynamicIncludeFlatBuilder);
+            }
+            assertFalse("a".equals(key));
```

Comment:
At this level, "a" would not appear anyways, but I'd expect it to be filtered out inside "three"

Reply:
fixed